### PR TITLE
[WIP] Initial implementation

### PR DESCRIPTION
**Pull request by timkurvers**
This implementation is based on http://wowprogramming.com/docs and a good dose of trial and error. It's likely the widget hierarchy diagram on the aforementioned site incorrect/incomplete, out-of-date and/or not applicable to Wrath of the Lich King.

It's currently set up to render either `FrameXML`, `GlueXML` or `Blizzard_BarbershopUI`. It does so by loading up the corresponding `.toc`. These table of contents files contain various `.xml` and `.lua` references, which are loaded sequentially (to ensure proper load order) through a core UI instance. This core instance instantiates components (called widgets) as well as Lua scripts. It keeps an atlas of all components by name. Every single component receives a reference to this core instance so that components may reference other components globally.

For Lua support, this PR is currently using [fengari]. To get [cheerio] (for parsing XML) and [fengari] to play nice in combination with webpack, we need to configure `typeof process` such that it meets requirements of both libraries.

[fengari] checks for [existence of `process`](https://github.com/fengari-lua/fengari/blob/master/src/loslib.js#L208-L215):

```javascript
if (typeof process === "undefined") {
    syslib.clock = function(L) {
        lua_pushnumber(L, performance.now()/1000);
        return 1;
    };
} else {
    /* Only with Node */
    const fs = require('fs');
    ...
}
```

Whereas `process-nextick-args` (an indirect dependency of [cheerio] via `htmlparser2` and `readable-stream`) on the other hand, [assumes `process` is present](https://github.com/calvinmetcalf/process-nextick-args/blob/master/index.js#L3-L9):

```javascript
if (!process.version ||
    process.version.indexOf('v0.') === 0 ||
    process.version.indexOf('v1.') === 0 && process.version.indexOf('v1.8.') !== 0) {
  module.exports = nextTick;
} else {
  module.exports = process.nextTick;
}
```
 
TODOs:

- [ ] Set up an update/render loop (currently only renders on page load and/or window resize)
- [ ] Parse anchor points (most important bit to get proper visual rendering)
- [x] Solve the [fengari] / [cheerio] issue (see above)
- [ ] Implement [Widget API](http://wowprogramming.com/docs/widgets)
- [ ] Implement [Lua API](http://wowprogramming.com/docs/api)
- [ ] Implement [events](http://wowprogramming.com/docs/events)
- [ ] Performance (loading FrameXML seems to take an eternity right now, and that's without Lua execution)

[cheerio]: https://github.com/cheeriojs/cheerio
[fengari]: https://github.com/fengari-lua/fengari

### CODE REVIEW

File: src/lib/components/ScriptObject.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    this._parent = parent;
+
+    if (this._parent) {
+      this._parent.frames.add(this);
+    }
+  }
+
+  // LUA interface
+  GetName() {
+    return this.name;
```
Comment (fallenoak) [inline src/lib/components/ScriptObject.js]: How is `this` determined for the Lua-side code? In the client, when `CScriptObject_GetName` is called, this logic fires off:

```cpp
CScriptObject* object = FrameScript_GetObjectThis(L, CScriptObject::s_objectType);
```

where `L` is `lua_State` and `object` is the object `GetName` is being called upon.

File: src/lib/components/Region.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import ScriptRegion from './ScriptRegion';
+
+class Region extends ScriptRegion {
```
Comment (fallenoak) [inline src/lib/components/Region.js]: Let's rename this to `SimpleRegion` to match the client's naming scheme.

File: src/lib/components/ScriptRegion.js [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+  }
+
+  loadXML(node) {
+    super.loadXML(node);
+
+    const [width, height] = Dimension.parse(node.find('> Size'));
+    this.explicitWidth = width;
+    this.explicitHeight = height;
+
+    node.find('> Anchors > Anchor').each((index, element) => {
```
Comment (fallenoak) [inline src/lib/components/ScriptRegion.js]: Have we tried `node.children('Anchors').find('> Anchor')` instead? Not sure how Cheerio performance is holding up, but I see you did switch to `children()` elsewhere.

File: src/lib/lua/utils.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+      case LUA_TSTRING:
+        value = lua_tojsstring(state, idx);
+        break;
+      case LUA_TUSERDATA:
+        value = checkjs(state, idx);
+        break;
+      case LUA_TFUNCTION:
+        value = lua_tocfunction(state, idx);
+        if (!value) {
+          value = function LuaClosure() {};
```
Comment (fallenoak) [inline src/lib/lua/utils.js]: `function LuaClosure() {}` should probably be a static somewhere.

File: src/lib/lua/utils.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+const stackToArray = (state) => {
+  const top = lua_gettop(state);
+
+  const results = [];
+  for (let idx = 1; idx <= top; ++idx) {
+    const type = lua_type(state, idx);
+    let value;
+    switch (lua_type(state, idx)) {
+      case LUA_TNONE:
```
Comment (fallenoak) [inline src/lib/lua/utils.js]: Is this `case` supposed to fall through to `LUA_TNIL`?

File: src/lib/widgets/Frame.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
+import ScriptRegion from '../components/ScriptRegion';
+
+class Frame extends ScriptRegion {
```
Comment (fallenoak) [inline src/lib/widgets/Frame.js]: Any reason we shouldn't call this `SimpleFrame` to match the client?

File: src/lib/components/ScriptObject.js (hunk 1 above)
Comment (timkurvers) [reply src/lib/components/ScriptObject.js]: `this` is currently the widget `GetName` is called upon. Example from the achievement UI addon:

```lua
function AchievementCategoryButton_OnLoad (button)
	...
	local buttonName = button:GetName();
	...
end
```

However, there may be scenarios where the LUA widget and the actual backing object are not the same but I haven't seen any instances of that yet.

File: src/lib/components/Region.js (hunk 2 above)
Comment (timkurvers) [reply src/lib/components/Region.js]: Let's change it, I'll leave a comment a bit further down about the split between widgets/components.

File: src/lib/components/ScriptRegion.js (hunk 3 above)
Comment (timkurvers) [reply src/lib/components/ScriptRegion.js]: We might be able to do `node.children('Anchors').children('Anchor')` but I'm not sure the end result is the same. Will give it a shot 👍

File: src/lib/lua/utils.js (hunk 4 above)
Comment (timkurvers) [reply src/lib/lua/utils.js]: Ah, good point 👍 The built in `tojs` function in `fengari-interop` comes up with a very strange representation of non-C-functions, so this is here for clarity.

File: src/lib/lua/utils.js (hunk 5 above)
Comment (timkurvers) [reply src/lib/lua/utils.js]: Unsure 😮  Gonna have to read up on `TNONE` and `TNIL`, I had assumed it was similar to `undefined` and `null`, which are the same when loosely compared in JavaScript.

Once the LUA interactions are clear, we may as well drop this entire debug utility.

File: src/lib/widgets/Frame.js (hunk 6 above)
Comment (timkurvers) [reply src/lib/widgets/Frame.js]: Let's go! 👍

**Comment by timkurvers** [comment]
> Separation between widgets and components

Yeah, this is a tricky one. I'm down with the `Simple` prefixes, although it will make it more complicated to go from a piece of LUA code (for example `button:GetName()` and find the associated client implementation.

I can move the ones that are not implicitly used in the UI into an `abstract` folder. Should we call them widgets or components?

> Bitwise operators:

According to the [LUA 5.3 manual](https://www.lua.org/manual/5.3/manual.html):

> The bit32 library has been deprecated. It is easy to require a compatible external library or, better yet, to replace its functions with appropriate bitwise operations. (Keep in mind that bit32 operates on 32-bit integers, while the bitwise operators in Lua 5.3 operate on Lua integers, which by default have 64 bits.)

Will do some poking around and see if there's anything we can use.

**Comment by fallenoak** [comment]
> Yeah, this is a tricky one. I'm down with the Simple prefixes, although it will make it more complicated to go from a piece of LUA code (for example button:GetName() and find the associated client implementation.

Hmm, I'm not quite sure why it'll be any harder. Is `fengari-interop` doing some magic to locate classes by name?

> Bitwise operators:

Whatever we use for bitwise operators, we're going to want to make sure the API exposed matches the one used by the Lua in the client. Potentially we could make this work by hooking up a compat layer a la `compat.lua`.

From what I've gathered, the client uses [Lua BitOp](http://bitop.luajit.org/), a C extension compiled into Lua at build time. The Lua BitOp docs should spell out the available functions and namespacing, etc. We probably want to expose all of it, since who knows what Addon authors may have used.

**Comment by timkurvers** [comment]
Sounds good on the bitwise operators 👍 

It's currently us doing the lookups: https://github.com/wowserhq/ui/pull/1/files#diff-f6594fc1384e6f1b5c2cb1076cbcf67bR27.

Calls like `CreateFrame` and `createFrameFromXML` (through XML) will get routed through there, and the right implementation is picked. There's a one-to-one correspondence between the actual objects in LUA and the JS ones, `fengari-interop` makes this possible. For simplicity sake `GetName` is defined directly on `Button` / `SimpleButton` or any of its superclasses, rather than pointing to a separate function, which I think is how the actual client does it?

I think there might be some cognitive overhead as LUA/XML in the client does not use the `Simple` prefix, but that's probably doable.

**Comment by timkurvers** [comment]
Turns out the issue where certain widget names get wrongly suffixed with a floating point rather than integers – such as `Button1.0` instead of `Button1`, is a combination of behaviour in `fengari-interop` and how recent versions of Lua interpolate floats into strings.

Have opened an issue https://github.com/fengari-lua/fengari-interop/issues/31.

**Comment by timkurvers** [comment]
Superceded by https://github.com/wowserhq/client/pull/1.

**Pull request closed without merging**